### PR TITLE
AI SDK testing and ARM template-less projec tcreation.

### PR DESCRIPTION
**Pull request by MilesHolland**
A fork of this branch https://github.com/ZhidaLiu/azure-sdk-for-python/tree/user/zhili/sdkDefaultRG

The purpose is to account for cases where someone who only has the developer role for an AI resource/workspace hub tries to make a project/lean workspace beneath it. The developer role technically provides sufficient permissions to do this, but since normal workspace creation flow in the SDK requires ARM template deployment permissions (which the dev role doesn't have), the attempt fails.

This PR adds a check to see if a workspace creation request was for a project and failed due to permissions issues. If that's the case it tries again using the workspace dataplane API (whose restclient files are also added in this PR).

Also included are a bunch of e2e tests, including some tests that target this new behavior. Other tests include AI resource and connection tests in the AI SDK.

### CODE REVIEW

File: sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import pytest
+
+from azure.ai.resources.entities import AzureOpenAIConnection, AzureAISearchConnection, AzureAIServiceConnection
+from azure.ai.ml._restclient.v2023_06_01_preview.models import ConnectionAuthType
+from azure.ai.ml._utils.utils import camel_to_snake
+from azure.ai.ml.entities._credentials import ApiKeyConfiguration
+from azure.core.exceptions import ResourceNotFoundError
+
+
+@pytest.mark.e2etest
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py:14]:
```suggestion
@pytest.mark.unittest
```

File: sdk/ai/azure-ai-resources/tests/projects/e2etests/test_project_operations.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+from azure.core.exceptions import ResourceNotFoundError
+
+
+# Makes the following setup assumptions:
+# 2 projects already exist in the conftest's ai_client:
+# e2e_test_proj_1 and e2e_test_proj_2.
+# The former is the client's default project.
+
+@pytest.mark.e2etest
+class TestProjects:
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/tests/projects/e2etests/test_project_operations.py:16]:
```suggestion
@pytest.mark.usefixtures("recorded_test")
class TestProjects:
```

**Review by kdestin — CHANGES_REQUESTED** [review]

File: sdk/ai/azure-ai-resources/conftest.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        credential.get_token("https://management.azure.com/.default")
+    except Exception:
+        try:
+            credential = DefaultAzureCredential(process_timeout=60)
+            # Check if given credential can get token successfully.
+            credential.get_token("https://management.azure.com/.default")
+        except Exception:
+            credential = InteractiveBrowserCredential()
+
+    return credential
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/conftest.py:205]:
```suggestion
```

Appears to be a duplicate of the credential above it

File: sdk/ai/azure-ai-resources/tests/projects/e2etests/test_project_operations.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        poller_1 = ai_client_with_ai_access.projects.begin_create(project=new_local_project)
+        new_local_project = Project(
+            name="e2e_test_proj_2",
+            ai_resource="/subscriptions/b17253fa-f327-42d6-9686-f3e553e24763/resourceGroups/hanchi-test/providers/Microsoft.MachineLearningServices/workspaces/e2e_test_res_1",
+            description="e2e test project. DO NOT DELETE",
+        )
+        poller_2 = ai_client_with_ai_access.projects.begin_create(project=new_local_project)
+        poller_1.result()
+        poller_2.result()
+    '''
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/tests/projects/e2etests/test_project_operations.py:34]: Appears to be dead code?

File: sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    def test_aoai_entity(self):
+        pass
+
+
+    def test_search_entity(self):
+        pass
+
+    
+    def test_service_entity(self):
+        pass
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py:28]: Appears to be dead code 

```suggestion
```

File: sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ 
+    # TODO move to unit tests
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py:30]: TODO was accomplished

```suggestion
```

File: sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    def test_set_keys(self):
```
Comment (kdestin) [inline sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py:31]: Test makes no assertions. Unless the goal is that it doesn't throw an exception?

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_hub_operation.py [hunk 8]
```diff
@@ -19,6 +19,9 @@
 from azure.core.polling import LROPoller
 from azure.core.tracing.decorator import distributed_trace
 
+from azure.ai.ml.constants._common import Scope, ArmConstants
+from azure.ai.ml.entities._workspace_hub.workspace_hub import WorkspaceHub
+from azure.ai.ml.entities._workspace_hub._constants import WORKSPACE_HUB_KIND
```
Comment (luigiw) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_hub_operation.py:24]: Nit-pick: Are these needed?

File: sdk/ai/azure-ai-resources/azure/ai/resources/operations/_connection_operations.py [hunk 9]
```diff
@@ -60,14 +60,14 @@ def get(self, name: str, **kwargs) -> BaseConnection:
 
         # It's by design that both API and V2 SDK don't include the secrets from API response, the following
         # code fills the gap
+        # TODO do we still want this?
```
Comment (luigiw) [inline sdk/ai/azure-ai-resources/azure/ai/resources/operations/_connection_operations.py:63]: Revert this?

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations.py [hunk 10]
```diff
@@ -331,3 +362,47 @@ def callback(_, deserialized, args):
         poller = self._operation.begin_diagnose(resource_group, name, parameters, polling=True, cls=callback)
         module_logger.info("Diagnose request initiated for workspace: %s\n", name)
         return poller
+
+    @distributed_trace
+    def begin_join(self, workspace: Workspace, **kwargs: Dict) -> LROPoller[Workspace]:
```
Comment (luigiw) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_workspace_operations.py:367]: Let's make this a private function to simplify the API view conversation.

**Review by luigiw — CHANGES_REQUESTED** [review]

File: sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py (hunk 1 above)
Comment (MilesHolland) [reply sdk/ai/azure-ai-resources/tests/connections/unittests/test_connection_entity.py]: remove this test file for now

File: sdk/ai/azure-ai-resources/tests/projects/e2etests/test_project_operations.py (hunk 4 above)
Comment (MilesHolland) [reply sdk/ai/azure-ai-resources/tests/projects/e2etests/test_project_operations.py]: Correct, removed.

File: sdk/ai/azure-ai-resources/azure/ai/resources/operations/_connection_operations.py (hunk 9 above)
Comment (MilesHolland) [reply sdk/ai/azure-ai-resources/azure/ai/resources/operations/_connection_operations.py:63]: Yes, as per our discussion about this behavior working again.

**Review by luigiw — APPROVED** [review]

**Pull request merged**